src/Opserver.Core/Data/Dashboard/Providers/SignalFxDataProvider.SignalFlow.cs

```diff
@@ -0,0 +1,805 @@
+            {
+                Channel = channel;
+                Program = program;
+                Start = start;
+                Stop = stop;
+                Resolution = (int)resolution.TotalMilliseconds;
+            }
+
+            public override string Type => "execute";
+            public string Channel { get; }
+            public string Program { get; }
+            public DateTime Start { get; }
+            public DateTime Stop { get; }
+            public int Resolution { get; }
+        }
+
+        private class ControlMessage : SignalFlowMessage, IExecutionResult
+        {
+            public override string Type => "control-message";
+            public string Channel { get; set; }
+            public ControlMessageEvent Event { get; set; }
+            [JsonPropertyName("timestampMs")]
+            public DateTime Timestamp { get; set; }
+            public int Progress { get; set; }
+        }
+
+        private class MetadataMessage : SignalFlowMessage, IExecutionResult
+        {
+            public override string Type => "metadata";
+            public string Channel { get; set; }
+            public MetadataProperties Properties { get; set; }
+            [JsonPropertyName("tsId")]
+            public string TimeSeriesId { get; set; }
+        }
+
+        private class MetadataProperties
+        {
+            public string Host { get; set; }
+            [JsonPropertyName("sf_originatingMetric")]
+            public string Metric { get; set; }
+            [JsonExtensionData]
+            public Dictionary<string, object> Tags { get; set; }
+
+        }
+
+        [StructLayout(LayoutKind.Explicit)]
+        private readonly struct DataMessageValue
+        {
+            public DataMessageValue(string timeSeriesId, double value, DataMessageValueType type)
+            {
+                Int32 = 0;
+                Int64 = 0;
+                Double = value;
+                TimeSeriesId = timeSeriesId;
+                Type = type;
+            }
+
+            public DataMessageValue(string timeSeriesId, uint value, DataMessageValueType type)
+            {
+                Double = 0;
+                Int64 = 0;
+                Int32 = value;
+                TimeSeriesId = timeSeriesId;
+                Type = type;
+            }
+
+            public DataMessageValue(string timeSeriesId, ulong value, DataMessageValueType type)
+            {
+                Double = 0;
+                Int32 = 0;
+                Int64 = value;
+                TimeSeriesId = timeSeriesId;
+                Type = type;
+            }
+
+            [FieldOffset(0)]
+            public readonly string TimeSeriesId;
+
+            [FieldOffset(8)]
+            public readonly uint Int32;
+            [FieldOffset(8)]
+            public readonly ulong Int64;
+            [FieldOffset(8)]
+            public readonly double Double;
+
+            [FieldOffset(16)]
+            public readonly DataMessageValueType Type;
```

Comment:
Given this is a determinate type size where uint to others may vary, should it go before the values? Or first even?

Reply:
It can't go at the beginning because the runtime complains about misalignment with a reference type, but it can definitely go before the values. Fixing!